examples/custom_diffusion/train.py

```diff
@@ -0,0 +1,1300 @@
+import json
+import requests
+import itertools
+from io import BytesIO
+
+import numpy as np
+import torch
+import torch.nn.functional as F
+import torch.utils.checkpoint
+import transformers
+from accelerate import Accelerator
+from accelerate.logging import get_logger
+from accelerate.utils import ProjectConfiguration, set_seed
+from huggingface_hub import create_repo
+from huggingface_hub import HfApi
+from packaging import version
+from PIL import Image
+from torch.utils.data import Dataset
+from torchvision import transforms
+from tqdm.auto import tqdm
+from transformers import AutoTokenizer, PretrainedConfig
+from clip_retrieval.clip_client import ClipClient
+
+import diffusers
+from diffusers import (
+    AutoencoderKL,
+    DDPMScheduler,
+    DiffusionPipeline,
+    DPMSolverMultistepScheduler,
+    UNet2DConditionModel,
+)
+
+from diffusers.models.cross_attention import CrossAttention
+from diffusers.optimization import get_scheduler
+from diffusers.utils import check_min_version, is_wandb_available
+from diffusers.utils.import_utils import is_xformers_available
+
+# sys.path.append('./')
+from model_pipeline import CustomDiffusionAttnProcessor, CustomDiffusionPipeline, set_use_memory_efficient_attention_xformers
+
+
+# Will error if the minimal version of diffusers is not installed. Remove at your own risks.
+check_min_version("0.14.0")
+
+logger = get_logger(__name__)
+
+
+def create_custom_diffusion(unet, freeze_model):
+    for name, params in unet.named_parameters():
+        if freeze_model == 'crossattn':
+            if 'attn2' in name:
+                params.requires_grad = True
+                print(name)
+            else:
+                params.requires_grad = False
+        elif freeze_model == "crossattn_kv":
+            if 'attn2.to_k' in name or 'attn2.to_v' in name:
+                params.requires_grad = True
+                print(name)
+            else:
+                params.requires_grad = False
+        else:
+            raise ValueError(
+                "freeze_model argument only supports crossattn_kv or crossattn"
+            )
+
+    # change attn class
+    def change_attn(unet):
+        for layer in unet.children():
+            if type(layer) == CrossAttention:
+                bound_method = set_use_memory_efficient_attention_xformers.__get__(layer, layer.__class__)
+                setattr(layer, 'set_use_memory_efficient_attention_xformers', bound_method)
+            else:
+                change_attn(layer)
+
+    change_attn(unet)
+    unet.set_attn_processor(CustomDiffusionAttnProcessor())
+    return unet
+
+
+def freeze_params(params):
+    for param in params:
+        param.requires_grad = False
+
+
+def retrieve(class_prompt, class_images_dir, num_class_images):
+    factor = 1.5
+    num_images = int(factor * num_class_images)
+    client = ClipClient(url="https://knn.laion.ai/knn-service", indice_name="laion_400m", num_images=num_images, aesthetic_weight=0.1)
```

Comment:
Could we maybe the user to this as a pre-step and assume that the class images have already been downloaded when starting training?

Reply:
Yes, I will remove it from the dependency and make it a separate pre-step option.